src/sink/nats.rs

```diff
@@ -0,0 +1,298 @@
+}
+
+impl offramp::Impl for Nats {
+    fn from_config(config: &Option<OpConfig>) -> Result<Box<dyn Offramp>> {
+        if let Some(config) = config {
+            let config: Config = Config::new(config)?;
+            let (dummy_tx, _) = bounded(1);
+            Ok(SinkManager::new_box(Self {
+                sink_url: TremorURL::from_offramp_id("nats")?,
+                config,
+                postprocessors: vec![],
+                reply_channel: dummy_tx,
+                connection: None,
+            }))
+        } else {
+            Err("Nats offramp requires a configuration.".into())
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Sink for Nats {
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        codec: &mut dyn Codec,
+        _codec_map: &HashMap<String, Box<dyn Codec>>,
+        mut event: Event,
+    ) -> ResultVec {
+        let ingest_ns = event.ingest_ns;
+        let insight = event.insight_ack();
+        if let Some(connection) = &mut self.connection {
+            for (value, meta) in event.value_meta_iter() {
+                let encoded = codec.encode(value)?;
+                let processed =
+                    postprocess(self.postprocessors.as_mut_slice(), ingest_ns, encoded)?;
+                let meta_data = meta.get_object("nats");
+                let mut headers = None;
+                let mut reply = None;
+                if let Some(meta_data) = meta_data {
+                    headers = meta_data.get("headers");
+                    reply = meta_data.get("reply");
+                }
+                for payload in processed {
+                    // prepare message reply
+                    let mut message_reply = None;
+                    if let Some(reply) = reply {
+                        if let Some(reply_str) = reply.as_str() {
+                            message_reply = Some(reply_str);
+                        }
+                    } else if let Some(config_reply) = &self.config.reply {
+                        message_reply = Some(config_reply.as_str());
+                    }
+                    // prepare message headers
+                    let mut key_val: Vec<(&str, &str)> = Vec::new();
+                    if let Some(headers) = headers {
+                        if let Some(headers_obj) = headers.as_object() {
+                            for (key, val) in headers_obj.iter() {
+                                if let Some(array) = val.as_array() {
+                                    for ele in array.iter() {
+                                        if let Some(ele) = ele.as_str() {
+                                            key_val.push((key, ele));
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    } else {
+                        for (key, val) in &self.config.headers {
```

Comment:
Should we always set those headers and then only overwrite them on the metadata?

Reply:
Right now, we use the headers from the metdata if they _exist_, otherwise we use the headers specified in the config. Isn't that what we should be doing?

Reply:
There are arguments for both, but the big one IMHO is that for rest, [where we also allow setting headers](https://github.com/tremor-rs/tremor-runtime/blob/main/src/sink/rest.rs#L813) in the config and via metadata, we do use the config headers + the metadata headers. So being consistent will help users to avoid unexpected behavior.

If we did an `either / or` in rest I'd say to stick to that but in the existing code we do a `and` so I think we should stick to this behavior.